File: packages/sdk/src/resources/films.ts
import type { TheatricalHTTPClient } from '../http/client';
import type { Film, FilmDetail, FilmFilter, FilmSearchFilter } from '../types/film';
import { filmSchema, filmDetailSchema } from '../types/film';
import { z } from 'zod';

/**
 * Films resource — now showing, coming soon, search, and detailed film information.
 *
 * Provides access to the Vista OCAPI film catalogue. Supports browsing
 * current and upcoming releases, searching with rich filters (genre, format,
 * language, rating, runtime), and retrieving full film detail including
 * cast, crew, and ratings from multiple sources.
 */
export class FilmsResource {
  constructor(private readonly http: TheatricalHTTPClient) {}

  /**
   * Parse and validate a raw film response from the API.
   * Throws a ZodError if the response doesn't match the expected shape.
   */
  private parseFilm(data: unknown): Film {
    return filmSchema.parse(data) as Film;
  }

  /**
   * Parse and validate an array of films from the API.
   */
  private parseFilms(data: unknown): Film[] {
    return z.array(filmSchema).parse(data) as Film[];
  }

  /**
   * Parse and validate a full film detail response.
   */
  private parseFilmDetail(data: unknown): FilmDetail {
    return filmDetailSchema.parse(data) as FilmDetail;
  }

  /**
   * List films currently showing at a given site (or across all sites).
   *
   * @param filters - Optional site filter
   * @returns Array of films currently in cinemas
   */
  async nowShowing(filters?: { siteId?: string }): Promise<Film[]> {
    const data = await this.http.get<unknown>('/ocapi/v1/films/now-showing', {
      params: filters as Record<string, string | number | boolean | undefined>,
    });
    return this.parseFilms(data);
  }

  /**
   * List films coming soon — announced but not yet in cinemas.
   *
   * @param filters - Optional site filter
   * @returns Array of upcoming films
   */
  async comingSoon(filters?: { siteId?: string }): Promise<Film[]> {
    const data = await this.http.get<unknown>('/ocapi/v1/films/coming-soon', {
      params: filters as Record<string, string | number | boolean | undefined>,
    });
    return this.parseFilms(data);
  }

  /**
   * Retrieve a film by its unique identifier.
   *
   * @param filmId - The UUID of the film
   * @returns The base film record
   */
  async get(filmId: string): Promise<Film> {
    const data = await this.http.get<unknown>(`/ocapi/v1/films/${filmId}`);
    return this.parseFilm(data);
  }

  /**
   * Retrieve full film detail including cast, crew, ratings, formats, and languages.
   *
   * This returns the extended `FilmDetail` type with nested crew arrays,
   * multiple rating sources (IMDB, Rotten Tomatoes, Metacritic), available
   * screening formats, and language options.
   *
   * @param filmId - The UUID of the film
   * @returns Full film detail with cast, crew, and ratings
   */
  async getDetail(filmId: string): Promise<FilmDetail> {
    const data = await this.http.get<unknown>(`/ocapi/v1/films/${filmId}/detail`);
    return this.parseFilmDetail(data);
  }

  /**
   * Search films with basic filters.
   *
   * @param filters - Genre, query string, site, and showing status filters
   * @returns Array of matching films
   */
  async search(filters: FilmFilter): Promise<Film[]> {
    const data = await this.http.get<unknown>('/ocapi/v1/films', {
      params: filters as Record<string, string | number | boolean | undefined>,
    });
    return this.parseFilms(data);
  }

  /**
   * Advanced film search with extended filters.
   *
   * Supports filtering by format (IMAX, 3D, Dolby Atmos), language,
   * rating classification, runtime range, release date range, and sorting.
   *
   * @param filters - Extended search filters including format, language, runtime range, sorting
   * @returns Array of matching films
   */
  async advancedSearch(filters: FilmSearchFilter): Promise<Film[]> {
    const params: Record<string, string> = {};

    if (filters.siteId) params.siteId = filters.siteId;
    if (filters.genre) params.genre = filters.genre;
    if (filters.query) params.query = filters.query;
    if (filters.nowShowing !== undefined) params.nowShowing = String(filters.nowShowing);
    if (filters.comingSoon !== undefined) params.comingSoon = String(filters.comingSoon);
    if (filters.limit !== undefined) params.limit = String(filters.limit);
    if (filters.offset !== undefined) params.offset = String(filters.offset);
    if (filters.ratingClassification) params.ratingClassification = filters.ratingClassification;
    if (filters.format) params.format = filters.format;
    if (filters.language) params.language = filters.language;
    if (filters.releaseDateFrom) params.releaseDateFrom = filters.releaseDateFrom;
    if (filters.releaseDateTo) params.releaseDateTo = filters.releaseDateTo;
    if (filters.minRuntime !== undefined) params.minRuntime = String(filters.minRuntime);
    if (filters.maxRuntime !== undefined) params.maxRuntime = String(filters.maxRuntime);
    if (filters.sortBy) params.sortBy = filters.sortBy;
    if (filters.sortOrder) params.sortOrder = filters.sortOrder;

    const data = await this.http.get<unknown>('/ocapi/v1/films/search', { params });
    return this.parseFilms(data);
  }
}

File: packages/sdk-py/theatrical/resources/films.py
"""Films resource — now showing, coming soon, search, and detailed film information."""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional

from pydantic import TypeAdapter

from theatrical.types.film import Film, FilmDetail, FilmFilter, FilmSearchFilter

if TYPE_CHECKING:
    from theatrical.http.client import TheatricalHttpProtocol

_film_list_adapter = TypeAdapter(list[Film])
_film_adapter = TypeAdapter(Film)
_film_detail_adapter = TypeAdapter(FilmDetail)


class FilmsResource:
    def __init__(self, http: TheatricalHttpProtocol) -> None:
        self._http = http

    async def now_showing(self, *, site_id: Optional[str] = None) -> list[Film]:
        params: dict[str, str] | None = {"siteId": site_id} if site_id else None
        raw = await self._http.get("/ocapi/v1/films/now-showing", params=params)
        return _film_list_adapter.validate_python(raw)

    async def coming_soon(self, *, site_id: Optional[str] = None) -> list[Film]:
        params: dict[str, str] | None = {"siteId": site_id} if site_id else None
        raw = await self._http.get("/ocapi/v1/films/coming-soon", params=params)
        return _film_list_adapter.validate_python(raw)

    async def get(self, film_id: str) -> Film:
        raw = await self._http.get(f"/ocapi/v1/films/{film_id}")
        return _film_adapter.validate_python(raw)

    async def get_detail(self, film_id: str) -> FilmDetail:
        raw = await self._http.get(f"/ocapi/v1/films/{film_id}/detail")
        return _film_detail_adapter.validate_python(raw)

    async def search(self, filters: FilmFilter) -> list[Film]:
        params: dict[str, str] = {}
        d = filters.model_dump(exclude_none=True)
        for key, value in d.items():
            params[key] = str(value) if not isinstance(value, str) else value
        raw = await self._http.get("/ocapi/v1/films", params=params or None)
        return _film_list_adapter.validate_python(raw)

    async def advanced_search(self, filters: FilmSearchFilter) -> list[Film]:
        params: dict[str, str] = {}
        d = filters.model_dump(exclude_none=True)
        for key, value in d.items():
            params[key] = str(value) if not isinstance(value, str) else value
        raw = await self._http.get("/ocapi/v1/films/search", params=params or None)
        return _film_list_adapter.validate_python(raw)
